server/src/main/java/io/spine/server/storage/RecordWithColumns.java

```diff
@@ -0,0 +1,211 @@
+ * @param <R>
+ *         the type of the stored records
+ */
+public class RecordWithColumns<I, R extends Message> {
+
+    private final I id;
+    private final R record;
+
+    /**
+     * A map of column names to the corresponding column values.
+     *
+     * @implNote It's impossible to use an {@link com.google.common.collect.ImmutableMap
+     *         ImmutableMap}, as the values may contain {@code null}s.
+     */
+    private final Map<ColumnName, @Nullable Object> storageFields;
+
+    protected RecordWithColumns(I identifier, R record, Map<ColumnName, Object> storageFields) {
+        this.id = checkNotNull(identifier);
+        this.record = checkNotNull(record);
+        this.storageFields = new HashMap<>(storageFields);
+    }
+
+    /**
+     * Creates a new record extracting the column values from the passed {@code Message} and setting
+     * the passed identifier value as the record identifier.
+     */
+    public static <I, R extends Message>
+    RecordWithColumns<I, R> create(I identifier, R record, RecordSpec<I, R, R> recordSpec) {
+        checkNotNull(identifier);
+        checkNotNull(record);
+        checkNotNull(recordSpec);
+        Map<ColumnName, @Nullable Object> storageFields = recordSpec.valuesIn(record);
+        return of(identifier, record, storageFields);
+    }
+
+    /**
+     * Creates a new record extracting the column values from the passed {@code Message}.
+     */
+    public static <I, R extends Message>
+    RecordWithColumns<I, R> create(R record, RecordSpec<I, R, R> recordSpec) {
+        checkNotNull(record);
+        checkNotNull(recordSpec);
+        Map<ColumnName, @Nullable Object> storageFields = recordSpec.valuesIn(record);
+        I identifier = recordSpec.idValueIn(record);
+        return of(identifier, record, storageFields);
+    }
+
+    /**
+     * Wraps a passed record.
+     *
+     * <p>Such instance of {@code RecordWithColumns} will contain no storage fields.
+     */
+    @Internal
+    public static <I, R extends Message> RecordWithColumns<I, R> of(I id, R record) {
+        return new RecordWithColumns<>(id, record, Collections.emptyMap());
+    }
+
+    /**
+     * Creates a new instance from the passed record and storage fields.
+     */
+    @VisibleForTesting
+    public static <I, R extends Message>
+    RecordWithColumns<I, R> of(I identifier, R record, Map<ColumnName, Object> storageFields) {
+        return new RecordWithColumns<>(identifier, record, storageFields);
+    }
+
+    /**
+     * Returns the identifier of the record.
+     */
+    public I id() {
+        return id;
+    }
+
+    /**
+     * Returns the message of the record.
+     */
+    public R record() {
+        return record;
+    }
+
+    /**
+     * Obtains the names of storage fields in the record.
+     *
+     * @return the storage field names
+     */
+    public ImmutableSet<ColumnName> columnNames() {
+        return ImmutableSet.copyOf(storageFields.keySet());
```

Comment:
Assuming, `storageFields` can be made immutable, its keys are going to be immutable too.